strawberryfields/tdm/tdmprogram.py

```diff
@@ -268,20 +386,149 @@ def __init__(self, N, name=None):
+            if device_gates != program_gates:
+                raise CircuitError(
+                    "The gates or the order of gates used in the Program is incompatible with the device '{}' ".format(
+                        device.target
+                    )
+                )
+
+            # Second check: the gates act on the correct modes
+            program_modes = [[r.ind for r in cmd.reg] for cmd in self.rolled_circuit]
+            device_modes = [op["modes"] for op in device_layout.operations]
+            if program_modes != device_modes:
+                raise CircuitError(
+                    "Program cannot be used with the device '{}' "
+                    "due to incompatible mode ordering.".format(device.target)
+                )
+
+            # Third check: the parameters of the gates are valid
+            # We will loop over the different operations in the device specification
+
+            for i, operation in enumerate(device_layout.operations):
+                # We obtain the name of the parameter(s)
+                param_names = operation["args"]
+
+                program_params_len = len(self.rolled_circuit[i].op.p)
+                device_params_len = len(param_names)
+                # The next if is to make sure we do not flag incorrectly things like Sgate(r,0) being different Sgate(r)
+                # This assumes that parameters other than the first one are zero if not explicitly stated.
+                if device_params_len < program_params_len:
+                    for j in range(1, program_params_len):
+                        if self.rolled_circuit[i].op.p[j] != 0:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter.".format(device.target)
+                            )
+                # Now we will check explicitly if the parameters in the program match
+                num_symbolic_param = 0  # counts the number of symbolic variables, which are labeled consecutively by the context method
+
+                for k, param_name in enumerate(param_names):
+                    # Obtain the value of the corresponding parameter in the program
+                    program_param = self.rolled_circuit[i].op.p[k]
+
+                    # make sure that hardcoded parameters in the device layout are correct
+                    if not isinstance(param_name, str):
+                        if not program_param == param_name:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter. Parameter has value '{}'"
+                                "while its valid value is '{}'".format(
+                                    device.target, program_param, param_name
+                                )
```

Comment:
Could these with the next two messages be factored out? There seems to be a minor difference